File: 0000-0999/0074. Search a 2D Matrix/Solution.py
'''
/**
 * Author: Fbin
 * Link:https://leetcode.com/problems/search-a-2d-matrix/description/?envType=problem-list-v2&envId=binary-search
 * Created at: 31/03/2025
 */
'''

from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        row, col = len(matrix), len(matrix[0]) if matrix else 0
        left, right = 0  , row * col - 1
        while(left <= right): 
            mid = left + (right - left) // 2
            mid_val = matrix[mid // col][mid % col]
            if mid_val == target: return True
            elif mid_val < target: left = mid + 1
            else: right = mid - 1
        return False

File: 0000-0999/0074. Search a 2D Matrix/Solution.cpp
/**
 * Author: Fbin
 * Link:https://leetcode.com/problems/search-a-2d-matrix/?envType=problem-list-v2&envId=binary-search
 * Created at: 31/03/2025
 */
#include <bits/stdc++.h>

using namespace std;

class Solution {
 public:
  bool searchMatrix(vector<vector<int>>& matrix, int target) {
    int row = matrix.size();
    int col = matrix[0].size();
    int l = 0;
    int r = row * col - 1;

    while (l <= r) {
      int m = l + (r - l) / 2;
      int value = matrix[m / col][m % col];
      if (value == target) return true;
      if (value < target) {
        l = m + 1;
      } else {
        r = m - 1;
      }
    }
    return false;
  }
};

// Solution at 26/05/2025
class Solution {
 public:
  bool searchMatrix(vector<vector<int>>& matrix, int target) {
    int m = matrix.size();
    int n = matrix[0].size();
    int l = 0, r = (m - 1) * n + (n - 1);
    while (l <= r) {
      int mid = l + (r - l) / 2;
      if (matrix[mid / n][mid % n] == target)
        return true;
      else if (matrix[mid / n][mid % n] < target)
        l = mid + 1;
      else
        r = mid - 1;
    }

    return false;
  }
};

// (0, 0) -> (m - 1, n - 1)
// 0 -> (m - 1) * n + (n - 1);

// m = 3, n = 4
// (0, 0) -> (2, 3)
// 0 -> 2 * 4 + 3 = 11
